File: python/day02/day02.py
from typing import List, Tuple
import sys

def parseInput() -> List[Tuple[str, int]]:
    result = []
    with open("day02/input.txt", 'r') as file:
        for line in file:
            direction_units = line.split(' ')
            t = (direction_units[0], int(direction_units[1]))
            result.append(t)
    return result

def calculateFinalPosition(input: List[Tuple[str, int]], with_aim: bool) -> Tuple[int, int]:
    horizontal = 0
    depth = 0
    aim = 0
    for direction, units in input:
        if direction == "forward":
            horizontal += units
            if with_aim:
                depth += (units*aim)
        elif direction == "down":
            if with_aim:
                aim += units
            else:
                depth += units
        elif direction == "up":
            if with_aim:
                aim -= units
            else:
                depth -= units
    return (horizontal, depth)

def part1(input: List[Tuple[str, int]]):
    final_position = calculateFinalPosition(input, False)
    return (final_position[0] * final_position[1])

def part2(input: List[Tuple[str, int]]):
    final_position = calculateFinalPosition(input, True)
    return (final_position[0] * final_position[1])

if __name__ == '__main__':
    input = parseInput()
    print(part1(input))
    print(part2(input))

File: typescript/day02/day02.ts
import * as fs from 'fs';
import * as path from 'path';

function readInput(filename: string): Command[] {
    let file = fs.readFileSync(filename, "utf-8")
    let commands: Command[] = [];

    file.split("\n").forEach((value) => {
        let [direction, unit] = value.split(' ');
        commands.push({direction: direction, units: parseInt(unit)})
    });

    return commands
}

interface Command {
    direction: string,
    units: number
}

interface Position {
    horizontal: number,
    depth: number,
    aim: number
}

function calculateFinalPosition(input: Command[], withAim: boolean): Position {
    const finalPosition: Position = { horizontal: 0, depth: 0, aim: 0};

    for (const command of input) {
        switch (command.direction) {
            case "forward":
                if (withAim) {
                    finalPosition.horizontal += command.units;
                    finalPosition.depth += (command.units * finalPosition.aim)
                } else {
                    finalPosition.horizontal += command.units;
                }
                break;
            case "down":
                if (withAim) {
                    finalPosition.aim += command.units;
                } else {
                    finalPosition.depth += command.units;
                }
                break;
            case "up":
                if (withAim) {
                    finalPosition.aim -= command.units;
                } else {
                    finalPosition.depth -= command.units;
                }
                break;
            default:
                break;
        }
    }

    return finalPosition
}

function main() {
    const filename = path.join(__dirname, '..', '..', 'inputs/day02.txt');
    const input = readInput(filename);

    const part1 = calculateFinalPosition(input, false);
    const part2 = calculateFinalPosition(input, true);
    console.log("Part 1: " + (part1.depth*part1.horizontal));
    console.log("Part 2: " + (part2.depth*part2.horizontal));
}

main()